---
title: "RnaSeqMetric plots"
author: "Chris Stubben"
output: html_document
params:
   output: Metrics
   pattern: txt$
   recursive: TRUE
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, comment="")
opts <- paste0('pattern="', params$pattern, '", path="', params$output, '"')
if(params$recursive)  opts <- paste0(opts, ", recursive=TRUE")
library(hciR)
library(highcharter)
library(tidyr)
library(DT)
rna <- read_RnaSeqMetrics(path=params$output, pattern=params$pattern, recursive=params$recursive)
```

This file was created using [R Markdown] and reads output files from [CollectRnaSeqMetrics].
The [link] to definitions on the [CollectRnaSeqMetrics] page describes the different metrics in more detail.

### Transcript coverage

```{r cov, echo=FALSE}
hchart(rna$coverage, "line", x=Position, y=Coverage, group=Sample) %>%
 hc_plotOptions(series=list(showInLegend=FALSE)) %>%
  hc_tooltip(pointFormat = "<b>{series.name}</b><br>{point.x}, {point.y:.3f}", headerFormat="") %>%
   hc_chart(zoomType = "xy")  %>%
    hc_exporting(enabled=TRUE, filename = "coverage")
```

### Percent aligned

<br>

```{r metrics, echo=FALSE}
x <- spread(rna$metrics, Key, Value)
x <- x[, c("Sample", "Pf_aligned_bases", "Pct_coding_bases", "Pct_utr_bases", "Pct_intergenic_bases",
           "Pct_intronic_bases", "Pct_ribosomal_bases", "Pct_usable_bases")]
names(x) <- gsub("Pct_|_bases", "",  names(x))
names(x)[1:2] <- c("sample", "aligned bases")
### turn off pagination for small tables ONLY  where dom = "p"
DOM <- "Brtip"
if(nrow(x)<=10) DOM <- "Brti"
datatable(x,  extensions = 'Buttons', options = list(
    dom = DOM,
    buttons = c('copy', 'csv', 'excel', 'pdf')
  )) %>% formatPercentage( 3:8, 2)
```

<br>


[highcharter]: http://jkunst.com/highcharter/index.html
[R Markdown]: http://rmarkdown.rstudio.com
[CollectRnaSeqMetrics]: https://broadinstitute.github.io/picard/command-line-overview.html#CollectRnaSeqMetrics
[link]: https://broadinstitute.github.io/picard/picard-metric-definitions.html#RnaSeqMetrics
