---
title: "DESeq analysis"
output: html_document
date: '`r gsub("  ", " ", format(Sys.time(), "%B %e, %Y"))`'
params:
   samples: samples.txt
   counts: counts.txt
   database: human
   filter: 5
   padj: 0.05
   trt: trt
   vs: NA
   relevel: NA
   mouseover: NA
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, comment="# ", collapse=TRUE)
library(hciR)
## replace params$... in code blocks.
## See http://stackoverflow.com/questions/43699235/replace-variables-with-their-corresponding-values-in-source-code-chunk
library(knitr)
library(readr)
default_source_hook <- knit_hooks$get('source')
knit_hooks$set(source = function(x, options) {
  x <- gsub('params\\$samples',  paste0('"',params$samples,'"'), x)
  x <- gsub('params\\$counts',   paste0('"',params$counts,'"'), x)
  x <- gsub('params\\$database', paste0('"',params$database,'"'), x)
  x <- gsub('params\\$filter', params$filter, x)
  x <- gsub('params\\$padj',   params$padj, x)
  default_source_hook(x, options)
})
```

This guide follows the [Bioconductor RNA-Seq workflow] to find differentially expressed
genes using [DESeq2] and the [hciR] package on Github to simplify the [R] code.

### Load samples and counts

Load samples with ids matching count column names and treatment column for contrasts.

```{r load_samples, message=FALSE, echo=2:3}
options(width=110)
samples <- read_tsv(params$samples)
samples
```

```{r relevel, echo=FALSE}
if(params$relevel != "NA"){
   lvls <- strsplit( params$relevel, ", *")[[1]]
   dds$trt <- factor(dds$trt, levels = lvls)
}
```


```{r check_samples, include=FALSE}
## check other names like Trt, treatment
n <- grep("Trt|[Tt]reat", colnames(samples))
if(length(n) == 1)  colnames(samples)[n] <- "trt"
if(!"trt" %in% colnames(samples)) stop("A trt column is required for the DESeq design formula")
## add option for vst?
if(nrow(samples)> 100) stop("Too many samples for rlog, try using vst")
```

Load the count matrix.

```{r load_counts, message=FALSE, echo=2:3}
options(width=110)
counts <- read_tsv(params$counts)
counts
```

Remove features with zero counts and features with fewer than `r params$filter` reads
in any sample.


```{r filter_counts}
counts <- filter_counts(counts, n = params$filter)
```

### Run DESeq

Combine samples and counts to run `DESeq` and get regularized log transforms (rlog)
for sample visualizations.

```{r run_DESeq_is_SLOW}
dds <- deseq_from_tibble(counts, samples, design = ~ params$trt)
rld <-  r_log(dds)
```

### PCA plot

Plot the first two principal components using the rlog values from the top 500
variable genes.  You can select a region and zoom into this interactive
highchart and hover over points to view sample ids.

```{r pcaplot1, eval=FALSE}
plot_pca(rld, "trt", width=700)
```

```{r pcaplot2, echo=FALSE}
if(params$mouseover != "NA"){
   tip <- strsplit( params$mouseover, ", *")[[1]]
   plot_pca(rld, "trt", tooltip=tip, width=700)
}else{
   plot_pca(rld, "trt", width=700)
}
```

### Sample distances

Cluster all the rlog values using the R function `dist` to calculate the Euclidean
distance between samples.

```{r plot_dist, fig.width=8, fig.height=6}
plot_dist(rld, "trt", border=NA)
```

### Gene annotations

Load the `r params$database` annotations from Ensembl.

```{r read_biomart, eval=FALSE}
genes <- read_biomart(params$database)
```

```{r read_biomart2, echo=FALSE}
## mmu has humna_homolog
if( params$database == "mouse"){
   data(mmu)
   genes <- mmu
}else if(params$database == "human"){
   data(hsa)
   genes <- hsa
}else{
   genes <- read_biomart(params$database)
}
```

### DESeq results

Get the DESeq results using a `r round(params$padj * 100)`% false discovery rate (FDR).

```{r results_block, eval=FALSE}
res <- results_all(dds, genes, alpha = params$padj)
```

```{r DESeq_results_with_lfcShrink_VERY_SLOW, echo=FALSE}
## Include human homolog with mouse results
if(params$database == "mouse"){
   if(params$vs == "NA"){
      res <- results_all(dds, genes,  simplify = FALSE, other="human_homolog")
   }else{
      res <- results_all(dds, genes, vs = params$vs, simplify = FALSE, other="human_homolog")
   }
}else{
   if(params$vs == "NA"){
     res <- results_all(dds, genes,  simplify = FALSE)
   }else{
     res <- results_all(dds, genes, vs = params$vs, simplify = FALSE)
   }
}
```


### Sample plots from the first contrast

Plot fold changes and p-values from protein coding genes in a volcano plot.


```{r volcano_block, eval=FALSE}
plot_volcano(res[[1]], ggplot=TRUE)
```

```{r volcano2, fig.width=6, fig.height=5, echo=FALSE}
plot_volcano(res[[1]], ggplot=TRUE, pvalue= -log10( sort(res[[1]]$padj)[20]))
```


Select the top 20 up- and down-regulated genes with the largest fold changes,
join to the rlog matrix  and cluster the rlog differences. The values in the
heatmap represent the amount a gene deviates in a specific sample from the
geneâ€™s average across all samples.

```{r gene_heatmap1, fig.width=7, fig.height=6}
x <- top_counts( res[[1]], rld, sort_fc = TRUE)
plot_genes(x, "trt", fontsize_row=8)
```
<br>



Select the top 500 genes sorted by p-value and scale by rows, so values now
represent the number of standard deviations from the mean rlog value.

```{r gene_heatmap2, fig.width=6.5, fig.height=6}
x <- top_counts( res[[1]], rld, top=500, padj=1)
plot_genes(x, "trt", scale="row", show_rownames=FALSE)
```
<br>

### Save results

Save the DESeq results to a single Excel file in `DESeq.xlsx` and R objects to a binary data file to load into a new session.
The  function will also output raw counts, rlog values, normalized counts, samples and gene annotations in the Excel file.

```{r write_results_to_Excel}
write_deseq(res, dds, rld, genes)
save(res, dds, rld, genes, file="dds.rda")
```



[R]: https://www.r-project.org
[Bioconductor RNA-Seq workflow]: http://www.bioconductor.org/help/workflows/rnaseqGene
[DESeq2]: http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html
[featureCounts]: http://bioinf.wehi.edu.au/featureCounts/
[hciR]: https://github.com/HuntsmanCancerInstitute/hciR
