---
title: "DESeq analysis for ZBTB33"
output: html_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, comment="# ", collapse=TRUE)
library(hciR)
```

This guide follows the [Bioconductor RNA-Seq workflow] to find differentially expressed genes
in the [GSE81784] dataset.

### Load samples and counts

Load the Sample Sheet from 13817R.

```{r load_samples, message=FALSE}
samples <- read_tsv("samples.txt")
samples
```
<br>

Load the nine `featureCount` output files in the `Counts` directory.

```{r load_counts, message=FALSE, echo=2:3}
options(width=120)
counts <- read_featureCounts("Counts/")
counts
```

Remove 24427 features with zero counts and 16705 features with fewer than 10 reads
in any sample to create a final count matrix with 16919 rows.

```{r filter_counts}
counts <- filter_counts(counts, n=10)
```

### Run DESeq

Combine the counts and samples to create a `DESeqDataSet` object and calculate
 the regularized log transform (rlog) for sample visualizations.


```{r DESeqobject}
dds <- deseq_from_tibble(counts, samples, design = ~trt)
rld <-  r_log(dds)
```


### PCA plot

Plot the first two principal components using the rlog values from the top 500 variable genes.
You can hover over points to view sample names or select a region to zoom into points in this interactive [highchart].

```{r pcaplot}
plot_pca(rld, "trt", tooltip=c("id", "name"), width=700)
```

### Sample distances

Cluster all the rlog values using the R function `dist` to calculate the Euclidean distance and plot sample distances.


```{r dist, echo=FALSE, fig.width=8, fig.height=6}
plot_dist(rld, "trt", border=NA)
```


### Gene annotations

Load the human annotations from Ensembl.

```{r biomaRt}
#hsa <- read_biomart("human")
data(hsa)
```


### DESeq results

Reorder the factor levels so scrambled control is last and then get the DESeq results
using a 5% false discovery rate (FDR).

```{r DESeq}
dds$trt <- factor(dds$trt, levels=c("si1", "si2", "sc"))
res <- results_all(dds, hsa)
```


### Comparing targeting siRNAs vs scrambled

Combine the siRNA treatments and compare to scrambled at at 1% FDR.

```{r DESeq_v2}
dds2 <- deseq_from_tibble(counts, samples, design = ~type)
rld2 <-  r_log(dds2)
dds2$trt <- factor(dds2$trt, levels=c("targeting", "scrambled"))
res2 <- results_all( dds2, hsa, alpha=0.01)
```


### Save results

Save the DESeq results to a single Excel file in DESeq.xlsx.  The  function will also output
raw counts, rlog values, normalized counts, samples and gene annotations.

```{r write, eval=FALSE}
write_deseq( dds2, res2, rld2, hsa)
```


### Plot heatmaps

Select the top 40 genes sorted by p-value, join to the rlog matrix and cluster the rlog differences.
The values in the heatmap represent the amount a gene deviates in a specific sample from the geneâ€™s average
across all samples.

```{r pheatmap, fig.width=8, fig.height=7}
x <- top_counts( res2, rld2)
plot_genes(x, "type", fontsize_row=8)
```
<br>

<br>


[GSE81784]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81784
[R]: https://www.r-project.org
[Bioconductor RNA-Seq workflow]: http://www.bioconductor.org/help/workflows/rnaseqGene
[DESeq2]: http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html
[featureCounts]: http://bioinf.wehi.edu.au/featureCounts/
[hciR]: https://github.com/HuntsmanCancerInstitute/hciR
[highchart]: http://jkunst.com/highcharter/index.html
[pheatmap]: https://cran.r-project.org/web/packages/pheatmap/index.html
