---
title: "STAR alignments for `r params$run`"
output:
  html_document:
    highlight: pygments
params:
    run: missing
    analysis: missing
---

```{r setup_GNomEx_links, include=FALSE}
##
usage <- 'render("README.Rmd", params=list(run="14011R1", analysis="A4124"))'
if(params$analysis == "missing") stop("An analysis ID is needed for GNomEx file links.  Usage: ", usage)
if(params$run == "missing")      stop("A run ID is needed for the report title.  Usage: ", usage)
analysis_dir <- paste0( "DownloadAnalysisSingleFileServlet.gx?idAnalysis=",
                      substring(params$analysis, 2),
                     "&view=Y&fileName=/Repository/AnalysisData/2017/", params$analysis)
```

This guide describes the steps used to run [STAR] alignments on the FASTQ files in GNomEx.

###1. Prepare reference sequence

Download the rat FASTA and GTF file from the [Ensembl] release 87 and run [STAR]
with the `genomeGenerate` option to create the reference database.  The
`sjdbGTFfile` option extracts splice junctions from the GTF file with a maximum
possible overhang of 49 bases (for 50 bp reads).

```{bash, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-87/fasta/rattus_norvegicus/dna/Rattus_norvegicus.Rnor_6.0.dna.toplevel.fa.gz
wget ftp://ftp.ensembl.org/pub/release-87/gtf/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.87.gtf.gz
gunzip *.gz

STAR --runMode genomeGenerate  --genomeDir star50 \
--genomeFastaFiles Rattus_norvegicus.Rnor_6.0.dna.toplevel.fa --runThreadN 12 \
--sjdbGTFfile Rattus_norvegicus.Rnor_6.0.87.gtf --sjdbOverhang 49 &
```

###2. Run alignments

Create a [cmd.txt] file to run alignments on the [CHPC] clusters following the
[Pysano] reference manual.  This file includes [shell] commands to parse sample
names from each FASTQ file and then read and write to different output files
with that prefix. For a single sample without all the variable substitution, the
STAR command is displayed below.  In this example, STAR reads the 13455X1 FASTQ
file, trims the adapter sequence, aligns to the rat reference genome created
above, and outputs a BAM file sorted by coordinates and bedGraph file.

```{bash, eval = FALSE}
STAR --genomeDir star50 --readFilesIn 13455X1.fq --runThreadN 12 \
--clip3pAdapterSeq AGATCGGAAGAGCACACGTCTGAACTCCAGTCA \
--outSAMtype BAM SortedByCoordinate \
--outWigType bedGraph --outWigStrand Unstranded
```

[FeatureCounts] summarizes aligned reads overlapping features in the GTF file.
The default is to ignore multi-mapping reads or reads overlapping two or more
features, but these counts can be added using `-M --fraction` and `-O largestOverlap`
to count multi-mappers by fractions and assign reads to the
feature with the largest overlap. For the stranded Illumina single read
sequencing runs,  the reads will align in the opposite direction of the feature,
so the `-s 2` option is used to count reversely stranded reads. You can also
verify that by opening a BAM file in a genome viewer like [IGV].

```{bash, eval = FALSE}
featureCounts -T 12 -s 2 -a Rattus_norvegicus.Rnor_6.0.87.gtf -o 13455X1.counts 13455X1.bam
```

The [cmd.txt] file also includes quality control checks using [FastQC],
[samtools idxstats] and [CollectRnaSeqMetrics]. The final step in the file
converts bedGraph output from STAR to smaller bigWig files for viewing coverage
tracks in a genome browser like [IGV].

[Pysano] will transfer the FASTQ files and scripts in [cmd.txt] to the clusters
at [CHPC], execute the jobs and transfer the results back to the `Alignments`
directory on GNomEx.

###3. Check statistics

An [QC report] created by [MultiQC] includes interactive summary tables and
plots for all five command in the `cmd.txt` file.   The General
Statistics table at the top of the report includes summaries from FastQC like
total sequences (M Seqs), STAR (M Aligned), featureCounts (M Assigned) and
collectRNASeqMetrics (% of Aligned sequences mapping to CDS, UTR, Intronic and
Intergenic).   The remaining sections summarize outputs from each program.


###4. Differential Expression

See the [DESeq report] for further details on the differential expression analysis.

<br>

[RSEM]: https://github.com/deweylab/RSEM
[Ensembl]: http://uswest.ensembl.org/info/data/ftp/index.html
[IGV]: http://software.broadinstitute.org/software/igv/
[STAR]: https://github.com/alexdobin/STAR
[featureCounts]: http://bioinf.wehi.edu.au/featureCounts/
[FastQC]: http://www.bioinformatics.babraham.ac.uk/projects/fastqc/
[Pysano]: http://healthcare.utah.edu/huntsmancancerinstitute/research/shared-resources/center-managed/bioinformatics/pysano
[CHPC]: https://www.chpc.utah.edu/
[shell]: http://tldp.org/LDP/abs/html/index.html
[CollectRnaSeqMetrics]: https://broadinstitute.github.io/picard/command-line-overview.html#CollectRnaSeqMetrics
[samtools idxstats]: http://www.htslib.org/
[MultiQC]: http://multiqc.info/
[cmd.txt]: `r analysis_dir`/cmd.txt
[QC report]: `r analysis_dir`/multiqc_report.html
[DESeq report]: `r analysis_dir`/DESeq.html
