---
title: "Downloading TCGA datasets"
output: html_document
---


```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, comment="# ", collapse=TRUE)
```

Use the [TCGAbiolinks] package to download datasets from the [GDC].

```{r library, eval = FALSE}
library(TCGAbiolinks)
projects <- getGDCprojects()
names(projects)[8] <- "project"
```

```{r listprojects}
datatable(projects[, 7:9], width =700)
```
<br>
<br>

Details on searching the GDC are found on the [help pages].   Briefly, Legacy
data uses Entrez gene IDs and the hg19 reference.   Harmonized data use Ensembl
IDs and the hg38 reference. Some valid harmonized data inputs for `GDCquery` are
listed in the table below.



```{r harmonized, echo=FALSE, eval=FALSE}
x0 <- read_csv("https://docs.google.com/spreadsheets/d/1f98kFdj9mxVDc1dv4xTZdx8iWgUiDYO-qiFJINvmTZs/export?format=csv&gid=2046985454")
names(x0) <- c("data.category", "data.type", "workflow.type", "platform")
```

```{r dt, echo=FALSE}
datatable(x0, width=700)
```

<br>
<br>

Loading data from the GDC consists of three steps.  Creating a query, downloading data
and loading results into an R object (usually a BioConductor object like `SummarizedExperiment`).

The first step returns a list of files to download.  Note that `GDCquery` uses Harmonized data
by default (legacy = FALSE).


```{r query, eval=FALSE}
query <- GDCquery(project = "TCGA-BRCA",
         data.category = "Transcriptome Profiling",
             data.type = "Gene Expression Quantification",
         workflow.type = "HTSeq - Counts")
# --------------------------------------
# o GDCquery: Searching in GDC database
# --------------------------------------
# Genome of reference: hg38
# --------------------------------------------
# oo Accessing GDC. This might take a while...
# --------------------------------------------
# ooo Project: TCGA-BRCA
# --------------------
# oo Filtering results
# --------------------
# ooo By data.type
# ooo By workflow.type
# ----------------
# oo Checking data
# ----------------
# ooo Check if there are duplicated cases
# ooo Check if there results for the query
# -------------------
# o Preparing output
# -------------------
```


```{r query_results, echo=-1}
options(width=110)
x1 <- getResults(query)
dim(x1)
head(x1[, c(6:7, 3)])
table(as.character( x1$tissue.definition))
```

Download the 1222 files.  The files are saved into a GDCdata/ directory, so the function
also checks if files have already been downloaded.

```{r gdcdownload, eval=FALSE}
GDCdownload(query)
# Downloading data for project TCGA-BRCA
# GDCdownload will download 1222 files. A total of 310.760859 MB
# Downloading as: Fri_Dec_22_11_23_58_2017.tar.gz
# Downloading: 310 MB
```


Finally, load the data into R using `GDCprepare`.  This takes a while to run since it downloads
metadata and other details, so I usually save the R object to a file to load into a new session.

```{r gdcprepare, eval=FALSE}
brca <- GDCprepare(query)
# |=========================================================================| 100%    1 MB
# ...
# Starting to add information to samples
#  => Add clinical information to samples
# Add FFPE information. More information at:
# => https://cancergenome.nih.gov/cancersselected/biospeccriteria
# => http://gdac.broadinstitute.org/runs/sampleReports/latest/FPPP_FFPE_Cases.html
#  => Adding subtype information to samples
# Subtype information from:doi:10.1038/nature11412
# Downloading genome information (try:0) Using: Human genes (GRCh38.p10)
# Loading from disk
# From the 60483 genes we couldn't map 3448
# Replicates found.
# FFPE should be removed. You can do data with the following command:
# data <- data[,!data$is_ffpe]
#                              is_ffpe
# TCGA-A7-A26J-01A-11R-A277-07   FALSE
# TCGA-A7-A0DB-01A-11R-A00Z-07   FALSE
# ...
save(brca, file="brca_counts.rda")
```



The datset is a `SummarizedExperiment` object with row and column data.

```{r brca}
library(SummarizedExperiment)
brca
```

The row data contain gene IDs and names.

```{r}
genes <- as_tibble(as.data.frame(rowData(brca)))
genes
```

Join the table to the Ensembl annotations in the [hciR] pacakge to subset
counts by biotype like protein_coding.

```{r left_join}
library(hciR)
data(hsa)
genes <- left_join(genes, hsa, by=c(ensembl_gene_id="id"))
```


The column data includes 86 columns and many are empty or duplicates (four
state and updated columns). Some columns like treatment are actually a list of
`data.frames`, so it's difficult to subset and query the table.   I recently added
a `gdc_coldata` function to [hciR] package that removes many of these columns.   In this example, the
column data table has 28 fewer rows.  You can also summarize the columns using `summary_tibble`.


```{r coldata, echo=-1}
options(width=110)
dim(colData(brca))
meta <- gdc_coldata(brca)
dim(meta)
table(meta$definition)
table(meta$tumor_stage)
table(table(meta$patient))
x <- summary_tibble(meta)
datatable(x)
```

<br>
<br>

Finally, counts can be returned using the `assay` function.

```{r counts, echo=-1}
options(width=110)
brca_counts <- assay(brca)
dim(brca_counts)
brca_counts[1:4, 1:3]
```


If needed, you can subset the counts or the summarized experiment object.

```{r subset}
n1 <- which(genes$biotype == "protein_coding")
n2 <- which(meta$definition == "Primary solid Tumor")
tumor_counts <- brca_counts[n1, n2]
dim(tumor_counts)
# brca[n1, n2]
```



[hciR]: https://github.com/HuntsmanCancerInstitute/hciR
[GDC]: https://portal.gdc.cancer.gov/
[TCGAbiolinks]: http://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html
[help pages]: https://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/query.html
